### ISSUE TITLE
Feature Request: pre-exec hook

### ISSUE THREAD

**Issue by chillum** [PowerShell/PowerShell#15271]
## Summary of the new feature/enhancement

As a user I want to be able to display running app's name in window title.

I launch `kubectl port-forward ...` in one terminal tab and sometimes I'm completely lost like: do I have it still running? in which terminal?

## Proposed technical implementation details (optional)

In zsh I can do:

```sh
function precmd {
    print -Pn '\e]0;%1~\a'
}
function preexec {
    printf '\e]0;%s\a' $1
}
```

and this will set my title:
1. to the running command if one is running
2. to the path of the directory if shell is waiting

in PowerShell I can set the title to the directory name in the `prompt` function, but didn't find any analogue to `preexec`.

basically the idea is to have a hook function that will execute just before the command is executed, and this function will know what command line will be executed next. and then it will be possible to implement the desired customization.

**Comment by mklement0** [comment]
Something similar previously came up in  #14484, and my sense is that this would best be handled by introducing new event hooks exposed via `$ExecutionContext.InvokeCommand.*Action` properties, analogous to [`$ExecutionContext.InvokeCommand.PreCommandLookupAction`](https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.commandinvocationintrinsics.precommandlookupaction) / [`$ExecutionContext.InvokeCommand.PostCommandLookupAction`](https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.commandinvocationintrinsics.postcommandlookupaction):

For instance, `$ExecutionContext.InvokeCommand.PreExecutionAction` and `$ExecutionContext.InvokeCommand.PostExecutionAction`, which could receive the command line being executed as the event argument. This would have the added advantage that it would be a _host-independent_ mechanism; that is, it wouldn't be tied to consoles (terminals).

---

Here's a workaround using the current features; it relies on `PSReadLine`, which is the command-line editing module that is loaded by default.

```powershell
# Override the PSReadLine Enter key handler in order to inject
# custom logic just before submitting a command.
Set-PSReadLineKeyHandler Enter {

  # Get the text of the comamnd being submitted. 
  $line = $null; [Microsoft.PowerShell.PSConsoleReadLine]::GetBufferState([ref] $line, [ref] $null)

  if ($line.Trim() -and $line.Trim() -inotmatch '\[Console\]::Title\b|\$host\.ui\.RawUI\.WindowTitle\b') { # (Reasonably robustly) bypass title updating when the window title is explicitly set / queried.

    # Save the current window title in a global variable, to be used by the `prompt` function to restore it.
    # Note: On Unix, $host.ui.RawUI.WindowTitle only reports a value if one was previously assigned 
    #       to that property,  so we must fall back to a default.
    $global:__prevWinTitle = if ($host.ui.RawUI.WindowTitle) { $host.ui.RawUI.WindowTitle } else { 'pwsh' }
  
    # Update the window title to show the running command and when it was started.
    $host.ui.RawUI.WindowTitle = "{0} - started at: {1:T}" -f $line, [datetime]::Now

  }

  # Submit the command.
  [Microsoft.PowerShell.PSConsoleReadLine]::AcceptLine()
}

# Update the `prompt` function to restore the previous window title on completion of a command.
# Note that this prevents scripts from modifying the window title.
$function:prompt = "if (`$global:__prevWinTitle) { `$host.ui.RawUI.WindowTitle = `$global:__prevWinTitle; `$global:__prevWinTitle = '' }; $function:prompt"
```

**Comment by vexx32** [comment]
At a glance, it looks kinda like you could use the `PreCommandLookupAction` you mentioned for this purpose fairly easily:

```powershell
$ExecutionContext.InvokeCommand.PreCommandLookupAction = {
    param($command)

    if ($command -notmatch 'Set-StrictMode|PSConsoleHost') {
        $host.UI.RawUI.WindowTitle = "Running [$command]..."
    }
    else {
        $host.UI.RawUI.WindowTitle = "pwsh"
    }
}
```

I guess the question is -- is there something significantly different in a "true" pre-exec hook to what this offers? This hook runs slightly before a pre-exec hook would, but offers much the same thing, I think?

**Comment by mklement0** [comment]
@vexx32, these hooks are invoked _multiple_ times per command line (use `$ExecutionContext.InvokeCommand.PreCommandLookupAction = { write-verbose -vb pre }` to see the problem), which makes this workaround awkward, especially if you want to handle restoring the previous title correctly - see https://github.com/PowerShell/PowerShell/issues/14484#issuecomment-751221242.

**Comment by SeeminglyScience** [comment]
Instead of a key handler, you can override the function `PSConsoleHostReadLine` similar to how you would `profile`.

That combined with overriding `profile` itself provides hooks into both pre and post invocation.

**Comment by mklement0** [comment]
Yes, redefining `PSConsoleHostReadLine` is also an option, as you've demonstrated at https://github.com/PowerShell/PowerShell/issues/14484#issuecomment-751741668, but using `Set-PSReadLineKeyHandler Enter` strikes me as the cleaner option, because it uses an official extensibility mechanism.

(And you probably meant `prompt`, not `$profile`).

**Comment by daxian-dbw** [comment]
This feature request looks to be a duplicate of #14484 (Feature Request - "pre"prompt functionality).
